src/modules/emote_menu/stores/index.js

```diff
@@ -0,0 +1,176 @@
+import cdn from '../../../utils/cdn.js';
+
+export const COLOUM_COUNT = window.location.pathname.endsWith('/chat') ? 7 : 9;
+
+function chunkArray(array, size) {
+  if (array.length <= size) {
+    return [array];
+  }
+  return [array.slice(0, size), ...chunkArray(array.slice(size), size)];
+}
+
+const fuse = new Fuse([], {
+  keys: ['code'],
+  threshold: 0.25,
+});
+
+class EmoteStore extends SafeEventEmitter {
+  constructor() {
+    super();
+    this.colsCount = COLOUM_COUNT;
+
+    watcher.on('channel.updated', async () => {
+      await this.loadConstants();
+      this.load();
+      this.loadDependableEmotes();
+      this.createRows();
+    });
+  }
+
+  async loadConstants() {
+    if (this.constantProviders != null) return; // constants already been loaded
+
+    this.constants = new Map(); // non-channel dependent emotes
+
+    const twitchEmotes = await loadTwitchEmotes();
+    this.constantProviders = [...twitchEmotes, ...emojiCategories];
+
+    for (const {emotes: constantEmotes} of this.constantProviders) {
+      constantEmotes.forEach((emote) => this.constants.set(String(emote.id), emote));
+    }
+  }
+
+  load() {
+    this.providers = [
+      {
+        provider: {
+          id: 'bttv',
+          displayName: 'BetterTTV',
+          icon: Icons.IMAGE(cdn.url('/assets/logos/mascot.png'), 'BetterTTV'),
+        },
+        emotes: emotes.getEmotes(['ffz-channel', 'ffz-global', 'bttv-emoji']),
+      },
+      {
+        provider: {
+          id: 'ffz',
+          displayName: 'FrankerFaceZ',
+          icon: Icons.IMAGE(cdn.url('/assets/logos/ffz_logo.png'), 'FrankerFaceZ'),
+        },
+        emotes: emotes.getEmotes(['bttv-channel', 'bttv-personal', 'bttv', 'bttv-emoji']),
+      },
+    ];
+
+    this.emotes = this.constants;
+
+    for (const {emotes: providerEmotes} of this.providers) {
+      providerEmotes.forEach((emote) => this.emotes.set(String(emote.id), emote));
+    }
+
+    const collection = [...this.emotes.values()];
+    [this.defaultEmote] = collection;
+    fuse.setCollection(collection);
+  }
+
+  loadDependableEmotes() {
+    this.dependableProviders = [
+      {
+        provider: {
+          id: 'favorites',
+          displayName: 'Favorites',
+          icon: Icons.STAR,
+        },
+        emotes: emoteStorage
+          .getFavorites()
+          .map((id) => this.emotes.get(id))
+          .filter((emote) => emote != null),
+      },
+      {
+        provider: {
+          id: 'recents',
+          displayName: 'Frequently Used',
+          icon: Icons.CLOCK,
+        },
+        emotes: emoteStorage
+          .getFrecents()
+          .map((id) => this.emotes.get(id))
+          .filter((emote) => emote != null),
+      },
+    ];
+  }
+
+  setCols(cols) {
+    this.colsCount = cols;
+    this.createRows();
+  }
+
+  createRows() {
+    this.rows = [];
+    this.headers = [];
+
+    for (const {provider, emotes: providerEmotes} of [
+      ...this.dependableProviders,
```

Comment:
rather than having 3 kinds of providers, have you considered just having a single object for all providers and updating providers by key? why do they need to be separate?

Reply:
For the order they appear in the menu